acme/db/nosql/account.go

```diff
@@ -177,55 +178,106 @@ func (db *DB) CreateExternalAccountKey(ctx context.Context, name string) (*acme.
 	}
 
 	dbeak := &dbExternalAccountKey{
-		ID:        keyID,
-		Name:      name,
-		KeyBytes:  random,
-		CreatedAt: clock.Now(),
+		ID:              keyID,
+		ProvisionerName: provisionerName,
+		Name:            name,
+		KeyBytes:        random,
+		CreatedAt:       clock.Now(),
 	}
 
 	if err = db.save(ctx, keyID, dbeak, nil, "external_account_key", externalAccountKeyTable); err != nil {
 		return nil, err
 	}
 	return &acme.ExternalAccountKey{
-		ID:        dbeak.ID,
-		Name:      dbeak.Name,
-		AccountID: dbeak.AccountID,
-		KeyBytes:  dbeak.KeyBytes,
-		CreatedAt: dbeak.CreatedAt,
-		BoundAt:   dbeak.BoundAt,
+		ID:              dbeak.ID,
+		ProvisionerName: dbeak.ProvisionerName,
+		Name:            dbeak.Name,
+		AccountID:       dbeak.AccountID,
+		KeyBytes:        dbeak.KeyBytes,
+		CreatedAt:       dbeak.CreatedAt,
+		BoundAt:         dbeak.BoundAt,
 	}, nil
 }
 
 // GetExternalAccountKey retrieves an External Account Binding key by KeyID
-func (db *DB) GetExternalAccountKey(ctx context.Context, keyID string) (*acme.ExternalAccountKey, error) {
+func (db *DB) GetExternalAccountKey(ctx context.Context, provisionerName string, keyID string) (*acme.ExternalAccountKey, error) {
 	dbeak, err := db.getDBExternalAccountKey(ctx, keyID)
 	if err != nil {
 		return nil, err
 	}
 
+	if dbeak.ProvisionerName != provisionerName {
+		return nil, acme.NewError(acme.ErrorUnauthorizedType, "name of provisioner does not match provisioner for which the EAB key was created")
+	}
+
 	return &acme.ExternalAccountKey{
-		ID:        dbeak.ID,
-		Name:      dbeak.Name,
-		AccountID: dbeak.AccountID,
-		KeyBytes:  dbeak.KeyBytes,
-		CreatedAt: dbeak.CreatedAt,
-		BoundAt:   dbeak.BoundAt,
+		ID:              dbeak.ID,
+		ProvisionerName: dbeak.ProvisionerName,
+		Name:            dbeak.Name,
+		AccountID:       dbeak.AccountID,
+		KeyBytes:        dbeak.KeyBytes,
+		CreatedAt:       dbeak.CreatedAt,
+		BoundAt:         dbeak.BoundAt,
 	}, nil
 }
 
-func (db *DB) UpdateExternalAccountKey(ctx context.Context, eak *acme.ExternalAccountKey) error {
+func (db *DB) DeleteExternalAccountKey(ctx context.Context, keyID string) error {
+	_, err := db.db.Get(externalAccountKeyTable, []byte(keyID))
```

Comment:
I think we could do these together in an `db.db.Update` call. But I'm not sure if there's any actual value to that. The idea for that method was that it would be sort of atomic, but I doubt it actually is. Depends on the details of the database.

Reply:
In this case I thought it would make sense to be able to indicate that a key could not be found. Didn't know about `db.db.Update()` yet, but it seems that it might able to do return a similar error. Will take a look.